### ISSUE TITLE
Doesn't work in Next.js middleware (Edge functions)

### ISSUE THREAD

**Issue by will-stone** [webdiscus/ansis#13]
## Current behaviour

When using Ansis inside Next.js's `middleware.ts` file, colours aren't enabled.

## Expected behaviour

Colours shown in terminal when using Ansis in a Next.js edge function.

### Reproduction Example

[Codesandbox](https://codesandbox.io/p/sandbox/ansis-next-js-middleware-bug-9phq84?file=%2Fmiddleware.ts&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clofijxaw000c2e6b0kyjauua%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clofijxav00032e6bj9ql1dur%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clofijxav00092e6b50ms13n6%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clofijxav000b2e6b9iwx9ish%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B67.12583728999999%252C32.87416271000001%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clofijxav00032e6bj9ql1dur%2522%253A%257B%2522id%2522%253A%2522clofijxav00032e6bj9ql1dur%2522%252C%2522activeTabId%2522%253A%2522clofikjox007x2e6blfm7lj9w%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clofijxav00022e6bwfdtf11u%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A10%252C%2522endLineNumber%2522%253A10%252C%2522startColumn%2522%253A1%252C%2522endColumn%2522%253A5%257D%255D%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fmiddleware.ts%2522%252C%2522id%2522%253A%2522clofikjox007x2e6blfm7lj9w%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252F.devcontainer%252FDockerfile%2522%252C%2522id%2522%253A%2522clofirs2g01682e6bcchfipot%2522%252C%2522mode%2522%253A%2522temporary%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522clofijxav000b2e6b9iwx9ish%2522%253A%257B%2522id%2522%253A%2522clofijxav000b2e6b9iwx9ish%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clofijxav000a2e6bafp66wb5%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522path%2522%253A%2522%252F%2522%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clofit560020q2e6bf6ie46wl%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clofit560020q2e6bf6ie46wl%2522%257D%252C%2522clofijxav00092e6b50ms13n6%2522%253A%257B%2522id%2522%253A%2522clofijxav00092e6b50ms13n6%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522clofisjby01kq2e6betc9mo9c%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clofisjby01kq2e6betc9mo9c%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

## Environment

- OS: macOS
- version of Node.js: 18.16.0
- version of Webpack: Next.js uses Turbopack now.
- version of the Plugin: 1.5.6

## Additional context

I've narrowed this down to this line: https://github.com/webdiscus/ansis/blob/7fca74ca3c6289883710acc578943b7b0cb7075c/src/ansi-codes.js#L29 and `process.stdout` not being available in Next.js's middleware. Is the `stdout && 'isTTY' in stdout` necessary if you also require the `/^screen|^xterm|^vt100|^vt220|^rxvt|color|ansi|cygwin|linux/i.test(env.TERM)` to pass?

**Comment by will-stone** [comment]
PS thanks for this library 😄 It's probably the most versatile of the new fast colour loggers. Great to have a lib still supporting CJS too, unlike some other chalky libs 😬

**Comment by webdiscus** [comment]
Hello @will-stone,

thank you for the issue report!
I will fix it.

Of course, your PR would be welcome too ;-)

**Comment by will-stone** [comment]
I'm happy to PR it, for sure. However, I'm a little worried I could break other situations. Would you be happy for me to add a failing test PR instead? I can try to set up a test that mimics Next.js middleware environment.

**Comment by webdiscus** [comment]
> I can try to set up a test that mimics Next.js middleware environment.

Yes, please create a small repository with a reproducible problem. This will help me solve the issue faster.

**Comment by will-stone** [comment]
Oh sure, happy to do that instead too. The codesandbox can be a bit flaky. I'll let you know when I have a repo for you to spin up 👍

**Comment by will-stone** [comment]
Here you go: https://github.com/will-stone/nextjs-ansis

Thanks! 🙂

**Comment by webdiscus** [comment]
@will-stone 

how can be set runtime as `nodejs`? I need it for testing next.js in nodejs mode.

I set `export const runtime = 'nodejs'` in index.tsx or middelware.ts, not works,
env contains always: `NEXT_RUNTIME: 'edge'`

**Comment by will-stone** [comment]
Ah, middleware is always run in edge mode. This is the cause of the problem.

**Comment by webdiscus** [comment]
ok, no problem

**Comment by will-stone** [comment]
This is one of the reasons I've been looking at Ansis: other libraries rely on `node:tty` and node utils aren't available in edge runtime. Here's an example of that: https://github.com/jorgebucaran/colorette/issues/100 However, my investigation has also shown `process.stdout` isn't available either. Hoping we can work around that though.

There's a Next.js [issue here](https://github.com/vercel/next.js/issues/50033) that started me looking at this library. The author reckons Ansis works but maybe that was an older version.

**Comment by webdiscus** [comment]
I know, the `node:tty` not works in edge mode, therefore I don't use any node lib for it.
Locally by me, colors already works in edge mode. 
I must write many tests yet.

**Comment by webdiscus** [comment]
@will-stone 

can you please check the beta version `v1.6.0-beta.0`?
Thank you!

**Comment by will-stone** [comment]
It works! 

<img width="479" alt="CleanShot 2023-11-02 at 10 56 26@2x" src="https://github.com/webdiscus/ansis/assets/654103/96d99e48-06e8-4ac2-920c-8b0eaad9a6f6">

https://github.com/will-stone/nextjs-ansis/commit/85e413c35bf2c46b70a496c8dd47d30dff38d859

Thank you for this amazingly quick turnaround 🤯 This needs to be marketed to the Next.js peeps at Vercel as this lib could be their default for all logging now.

**Comment by webdiscus** [comment]
@will-stone Thank you for the suggestion.

> This needs to be marketed to the Next.js peeps at Vercel as this lib could be their default for all logging now.

Many developers would benefit from having a small/fast/powerful/universal color library.
And it would be nice if Next.js had ansis "out of the box".

I'm afraid that Next.js will reject Ansis due to its low popularity.
Ansis has 100k downloads per week and is used by 2.3k developers.

Although chalk is also developed and maintained by one person, it is a very popular person.

We can try to suggest Asisis in Next.js, but it should not be the author (me), but another developer. If I do this, it will be perceived as self-promotion/spam.

**Comment by will-stone** [comment]
Yeah sorry, more of a throw away comment really. I’ll put a thread in their GitHub discussions to direct people this way when looking for a color logger that supports all of Next.js 👍😊

**Comment by webdiscus** [comment]
@will-stone the support for Next.js is released in the `v2.0.0`.
## Note:
In the new release is changed named import:
NEW (`v2.x.x`) named import
```js
import ansis, { red } from 'ansis'; // default and named import
ansis.red`error`;
red`error`;
```
OLD (`v1.5.x`) named import:
```js
import ansis from 'ansis'; // default import
import { red } from 'ansis/colors'; // named import
ansis.red`error`;
red`error`;
```

**Comment by webdiscus** [comment]
can be closed the issue?

**Comment by will-stone** [comment]
Awesome! Thanks, yeah let's close and I'll open new issues if I find any 🙂
